### 1-3. Set Up Notebook

In [32]:
#import libraries
import pandas as pd
import numpy as np
import os

In [33]:
#create path
path = r'C:\Users\jacob\Documents\Data Analytics\Data Immersion\Achievement 4\10-2023 Instacart Basket Analysis'

In [34]:
#import customer data
cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

### 4. Wrangle Data

In [35]:
#wrangle columns (unneeded columns and name changes)
cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

All the columns have pertinent and unique data so we won't drop any of them. 

In [36]:
#change surnam column to last_name
cust.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [37]:
#change First Name to first_name
cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [38]:
#check output
cust.head()

user_id first_name last_name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [39]:
#change gender, age and state to both be all lower case letters
cust.rename(columns = {'Gender' : 'gender'}, inplace = True)
cust.rename(columns = {'STATE' : 'state'}, inplace = True)
cust.rename(columns = {'Age' : 'age'}, inplace = True)

In [40]:
#check output
cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [41]:
#change fam_status to marital_status
cust.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [42]:
#check output
cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

In [43]:
#check data types
cust.dtypes

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [44]:
# change data types for better memory
cust['user_id'] = cust['user_id'].astype('int32')
cust['gender'] = cust['gender'].astype('category')
cust['age'] = cust['age'].astype('int8')
cust['n_dependants'] = cust['n_dependants'].astype('int8')
cust['marital_status'] = cust['marital_status'].astype('category')
cust['income'] = cust['income'].astype('int32')

In [45]:
#check data types
cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   user_id         206209 non-null  int32   
 1   first_name      194950 non-null  object  
 2   last_name       206209 non-null  object  
 3   gender          206209 non-null  category
 4   state           206209 non-null  object  
 5   age             206209 non-null  int8    
 6   date_joined     206209 non-null  object  
 7   n_dependants    206209 non-null  int8    
 8   marital_status  206209 non-null  category
 9   income          206209 non-null  int32   
dtypes: category(2), int32(2), int8(2), object(4)
memory usage: 8.7+ MB


In [46]:
#check descriptive statistics
cust.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

A couple of things stand out. First, the average number of dependents looks reasonable at 1.49, however, I would expect the max number of children to be higher than 3. Could be accurate, but seems like a low max for a set of over 200,000 people. 

Second, the average income of customers seems high to me. The average income in the United States according to one source is right below 60,000. The US Census found the the median income in the US is just under 70,000. A couple reasons, first, this data only looks at 200,000 people, compared to the 330 million in the US.

Along with that the max income for this data is 593,901. This is really high and could be pulling the average higher. Could be a realistic number, but does seem like an outlier, especially because 75% of the incomes fall at or below 124,244.

### 5. Quality and Consistency Checks

In [47]:
#check for missing values
cust.isnull().sum()

user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

While there are a significant number of missing values in the first name column, it doesn't impact the data in any way. We shouldn't remove this data based on the fact that first names are missing.

The only way that this would impact the data is if a specific customer had two user_ids, one with their fisrt and last name, and one with only their last name. 

When checking for duplicates we should make sure to exclude first names so that we can see if any of the other data is a complete duplicate (except first name). That might indicate the customer has two accounts.

In [48]:
#check for duplicates
cust_dups = cust[cust.duplicated()]

In [49]:
cust_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

Based on this there are no full duplicates. We will run another check to see if there are any duplicates when we exclude first names.

In [50]:
#create a subset of the data where I drop first name column
cust_2 = cust.drop(columns = ['first_name'])

In [51]:
#check output
cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

In [52]:
#check output
cust_2.head()

user_id last_name  gender       state  age date_joined  n_dependants  \
0    26711  Esquivel  Female    Missouri   48    1/1/2017             3   
1    33890      Hart  Female  New Mexico   36    1/1/2017             0   
2    65803    Farley    Male       Idaho   35    1/1/2017             2   
3   125935     Hicks  Female        Iowa   40    1/1/2017             0   
4   130797   Gilmore  Female    Maryland   26    1/1/2017             1   

  marital_status  income  
0        married  165665  
1         single   59285  
2        married   99568  
3         single   42049  
4        married   40374

In [53]:
#check for duplicates in cust_2
cust_2_dups = cust_2[cust_2.duplicated()]

In [54]:
#check output
cust_2_dups

Empty DataFrame
Columns: [user_id, last_name, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

There are no duplicates with or with out the customer's first name. So we are good to proceed with the original data frame 'cust'

### 6. Combine Customer Data with Orders and Products Data

In [55]:
#import orders_procucts_merged data
ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [56]:
#check output ords_prods
ords_prods.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2398795        1             2            3            7   
2    473747        1             3            3           12   
3   2254736        1             4            4            7   
4    431534        1             5            4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge     busiest_day      busyness  \
0         Soda  ...    both  Regularly busy  Regular days   
1         Soda  ...    both  Regularly busy  Slowest days   
2         Soda  ...    both  Regularly busy  Slowest days   
3         Soda  ...    both      Least busy  Slowest days   
4         Soda  ...    both      Least busy  Slowest days   

  busiest_period_of_day max_order  loyalty_flag avg_price  spender_flag  \
0        Average orders        10  New Customer       NaN           NaN   
1        Average orders        10  New Customer  6.367535   Low Spender   
2           Most orders        10  New Customer  7.515725   Low Spender   
3        Average orders        10  New Customer  8.197565   Low Spender   
4           Most orders        10  New Customer  8.205295   Low Spender   

  median_days_since_prior_order         frequency_flag  
0                           NaN                    NaN  
1                          20.5  Non-frequent Customer  
2                          13.0       Regular Customer  
3                          10.0      Frequent Customer  
4                          20.0       Regular Customer  

[5 rows x 23 columns]

In [57]:
#check output of cust
cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

user_id is the column that we will use to merge ords_prods and cust dataframes

In [58]:
#check data types of ords_prods 'user_id'
ords_prods['user_id'].dtype

dtype('int32')

In [59]:
#check output
ords_prods['user_id'].dtype

dtype('int32')

In [60]:
#check shape of cust
cust.shape

(206209, 10)

In [61]:
#check shape of ords_prods
ords_prods.shape

(32404161, 23)

In [62]:
#check data types of ords_prods for memory purposes
ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32404161 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   order_id                       int32  
 1   user_id                        int32  
 2   order_number                   int8   
 3   day_of_week                    int8   
 4   hour_of_day                    int8   
 5   days_since_prior_order         float16
 6   product_id                     int32  
 7   add_to_cart_order              int32  
 8   reordered                      int8   
 9   product_name                   object 
 10  aisle_id                       int32  
 11  department_id                  int32  
 12  prices                         float64
 13  _merge                         object 
 14  busiest_day                    object 
 15  busyness                       object 
 16  busiest_period_of_day          object 
 17  max_order                      int8   
 18  loyal

In [63]:
#change data types for better memory usage
ords_prods['_merge'] = ords_prods['_merge'].astype('category')
ords_prods['prices'] = ords_prods['prices'].astype('float32')
ords_prods['busiest_day'] = ords_prods['busiest_day'].astype('category')
ords_prods['busyness'] = ords_prods['busyness'].astype('category')
ords_prods['busiest_period_of_day'] = ords_prods['busiest_period_of_day'].astype('category')
ords_prods['loyalty_flag'] = ords_prods['loyalty_flag'].astype('category')
ords_prods['spender_flag'] = ords_prods['spender_flag'].astype('category')
ords_prods['frequency_flag'] = ords_prods['frequency_flag'].astype('category')
ords_prods['_merge'] = ords_prods['_merge'].astype('category')

In [64]:
#check memory usage
ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32404161 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   order_id                       int32   
 1   user_id                        int32   
 2   order_number                   int8    
 3   day_of_week                    int8    
 4   hour_of_day                    int8    
 5   days_since_prior_order         float16 
 6   product_id                     int32   
 7   add_to_cart_order              int32   
 8   reordered                      int8    
 9   product_name                   object  
 10  aisle_id                       int32   
 11  department_id                  int32   
 12  prices                         float32 
 13  _merge                         category
 14  busiest_day                    category
 15  busyness                       category
 16  busiest_period_of_day          category
 17  max_order                     

unable to perform merge because there is already a merge column in ords_prods dataframe. I will drop that merge column and re run the code

I will also drop busiest_day column because a new column 'busyness' has been created which flags for more specific labels.

In [65]:
#permanently remove _merge column from ords_prods
ords_prods = ords_prods.drop(columns = ['_merge'])

In [66]:
#remove busiest_day from ords_prods
ords_prods = ords_prods.drop(columns = ['busiest_day'])

In [67]:
#check output
ords_prods.head()

order_id  user_id  order_number  day_of_week  hour_of_day  \
0   2539329        1             1            2            8   
1   2398795        1             2            3            7   
2    473747        1             3            3           12   
3   2254736        1             4            4            7   
4    431534        1             5            4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices      busyness  \
0         Soda  ...              7     9.0  Regular days   
1         Soda  ...              7     9.0  Slowest days   
2         Soda  ...              7     9.0  Slowest days   
3         Soda  ...              7     9.0  Slowest days   
4         Soda  ...              7     9.0  Slowest days   

  busiest_period_of_day max_order  loyalty_flag avg_price  spender_flag  \
0        Average orders        10  New Customer       NaN           NaN   
1        Average orders        10  New Customer  6.367535   Low Spender   
2           Most orders        10  New Customer  7.515725   Low Spender   
3        Average orders        10  New Customer  8.197565   Low Spender   
4           Most orders        10  New Customer  8.205295   Low Spender   

  median_days_since_prior_order         frequency_flag  
0                           NaN                    NaN  
1                          20.5  Non-frequent Customer  
2                          13.0       Regular Customer  
3                          10.0      Frequent Customer  
4                          20.0       Regular Customer  

[5 rows x 21 columns]

In [68]:
#merge cust and ords_prods
ords_prods_cust = ords_prods.merge(cust, on = 'user_id', indicator = True)

### 8. Export Data

In [69]:
ords_prods_cust.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))